Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,546 @@
+        return;
+    }
+    g_hasPostEnable = 1;
+    
+    kscm_memory_check_for_oom_in_previous_session();
+
+    if (g_isEnabled)
+    {
+        ksmemory_write_possible_oom();
+        g_memoryTracker = [[MemoryTracker alloc] init];
+    }
+}
+
+KSCrashMonitorAPI* kscm_memory_getAPI(void)
+{
+    static KSCrashMonitorAPI api =
+    {
+        .setEnabled = setEnabled,
+        .isEnabled = isEnabled,
+        .addContextualInfoToEvent = addContextualInfoToEvent,
+        .notifyPostSystemEnable = notifyPostSystemEnable,
+    };
+    return &api;
+}
+
+/**
+ Read the previous sessions memory data,
+ and unlinks the file to remove any trace of it.
+ */
+static void ksmemory_read(const char* path)
+{
+    int fd = open(path, O_RDONLY);
+    if (fd == -1) {
+        unlink(path);
+        return;
+    }
+    
+    size_t size = sizeof(KSCrash_Memory);
+    KSCrash_Memory memory = {};
+    if (!ksfu_readBytesFromFD(fd, (char *)&memory, (int)size)) {
+        close(fd);
+        unlink(path);
+        return;
+    }
+    
+    // get rid of the file, we don't want it anymore.
+    close(fd);
+    unlink(path);
+    
+    // validate some of the data before doing anything with it.
+    
+    // check the timestamp, let's say it's valid for the last week
+    // do we really want crash reports older than a week anyway??
+    const uint64_t kUS_in_day = 8.64e+10;
+    const uint64_t kUS_in_week = kUS_in_day * 7;
+    uint64_t now = ksdate_microseconds();
+    if (memory.timestamp <= 0 || memory.timestamp == INT64_MAX || memory.timestamp < now - kUS_in_week) {
+        return;
+    }
+    
+    // check pressure and level are in ranges
+    if (memory.level > KSCrashAppMemoryStateTerminal) {
+        return;
+    }
+    if (memory.pressure > KSCrashAppMemoryStateTerminal) {
+        return;
+    }
+    
+    // check app transition state
+    if (memory.state > KSCrashAppTransitionStateExiting) {
+        return;
+    }
+    
+    // Footprint and remaining should always = limit
+    if (memory.footprint + memory.remaining != memory.limit) {
+        return;
+    }
+    
+    // if we're at max, we likely overflowed or set a negative value,
+    // in any case, we're counting this as a possible error and bailing.
+    if (memory.footprint == UINT64_MAX) {
+        return;
+    }
+    if (memory.remaining == UINT64_MAX) {
+        return;
+    }
+    if (memory.limit == UINT64_MAX) {
+        return;
+    }
+
+    g_previousSessionMemory = memory;
+}
+
+/**
+ Mapping memory to a file on disk. This allows us to simply treat the location
+ in memory as a structure and the kernel will ensure it is on disk. This is also
+ crash resistant.
+ */
+static void ksmemory_map(const char* path)
+{
+    void *ptr = ksfu_mmap(path, sizeof(KSCrash_Memory));
+    if (!ptr) {
+        return;
+    }
+    
+    g_memory = (KSCrash_Memory *)ptr;
+    KSCrashAppMemory *memory = g_memoryTracker.memory;
+    
+    _ks_memory_update(^(KSCrash_Memory *mem) {
```

Comment:
Does it make sense to generalize this update code with the `_updateMappedMemoryFrom` method above?